### PR TITLE
Image upload to replace wheelmap classics photo upload

### PR DESCRIPTION
**Pull request by mutaphysis**
Added image upload API 
 - Added image upload POST method
 - Similar to file uploads
 - Stores images in separate collection
 - Allows linking to existing places
 
Added image listing API
 - GET api for querying images for a place
 - list of images with full s3 url

Added simple captcha service api
- API for querying captchas at /captcha.svg
  - returns svg content
- Captcha data is stored in db with the generated solution
- Old Captchas are automatically removed after some time via cron job
- Image uploads require non-outdated captcha
  - Matching same placeId, hashedIp, solution, appCode

Simple IP based throttling
 - Only for captcha & upload image api
 - Check timestamp & hashedIp for requests above a certain frequency
 - Could be generalised into something for every api request

### CODE REVIEW

**Comment by mutaphysis** [comment]
@opyh Some feedback would be welcome!

I definitely want to rename image -> photo.

File: both/api/images/images.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    type: String,
+  },
+  hashedIp: {
+    type: String,
+  },
+  appToken: {
+    type: String,
+  },
+  timestamp: {
+    type: Date,
```
Comment (opyh) [inline both/api/images/images.js]: Would like to have an ISO date in the JSON output – would you be okay with that?

File: server/image-upload.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+const shaHasher = new JSSHA('SHA3-256', 'TEXT');
+
+// TODO check actual mime type
+const allowedMimeTypes = ['image/png', 'image/jpg', 'image/jpeg', 'image/tiff', 'image/tif', 'image/gif'];
+
+function handleUploadRequest(req, res) {
+  try {
+    const query = url.parse(req.url, true).query;
+
+    const hashedIp = shaHasher.getHash('HEX', req.connection.remoteAddress);
```
[the remark this answers is not in the document]
Comment (mutaphysis) [reply server/image-upload.js]: Done

File: server/image-upload.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+function respondWithError(res, code, reason) {
+  console.error('Responding with error', code, reason);
+  respond(res, code, { error: { reason } });
+}
+
+// use non nsa sha3 family
+const shaHasher = new JSSHA('SHA3-256', 'TEXT');
+
+// TODO check actual mime type
+const allowedMimeTypes = ['image/png', 'image/jpg', 'image/jpeg', 'image/tiff', 'image/tif', 'image/gif'];
```
[the remark this answers is not in the document]
Comment (mutaphysis) [reply server/image-upload.js]: Done

File: server/image-upload.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+      mimeType,
+      remotePath: `place-images/${placeId}/${Random.secret()}${suffix ? `.${suffix}` : ''}`,
+      timestamp: new Date(),
+    };
+
+    console.log('Inserting image upload', attributes);
+    const _id = Images.insert(attributes);
+    const image = Images.findOne(_id);
+
+    Captchas.remove(usedCaptcha);
```
[the remark this answers is not in the document]
Comment (mutaphysis) [reply server/image-upload.js]: Done

File: server/image-upload.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+      return;
+    }
+
+    const throttled = shouldThrottleByIp(Images, hashedIp);
+    if (throttled) {
+      respondWithError(res, 429, 'Too many requests');
+      return;
+    }
+
+    const place = PlaceInfos.findOne(placeId);
```
[the remark this answers is not in the document]
Comment (mutaphysis) [reply server/image-upload.js]: Moved to task

File: both/api/images/images.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+import { PlaceInfos } from '../place-infos/place-infos';
+
+export const Images = new Mongo.Collection('Images');
+
+Images.schema = new SimpleSchema({
+  placeId: {
+    type: String,
+  },
+  hashedIp: {
```
[the remark this answers is not in the document]
Comment (mutaphysis) [reply both/api/images/images.js]: Done

File: both/api/images/images.js (hunk 1 above)
Comment (mutaphysis) [reply both/api/images/images.js]: Done

File: both/api/images/server/privileges.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+import { Images } from '../images';
+
+// Deny all client-side updates since we will be using methods to manage this collection
+Images.deny({
+  insert() { return true; },
+  update() { return true; },
+  remove() { return true; },
+});
+
+// TODO image apis are open to everyone - probably should be limited?
```
[the remark this answers is not in the document]
Comment (mutaphysis) [reply both/api/images/server/privileges.js]: 2 goes into separate commit

**Pull request merged**